### DISCUSSION TITLE
Apparently "robot window" always shows up in a separate browser window?

### DISCUSSION THREAD

**Discussion by amrytov** [cyberbotics/webots#5423] [Q&A]
Hello,

Webots documentation tells us that "robot windows" would show up in a special panel in Webots UI.  I distinctly remember running the robot windows samples (but probably in an older WeBots version) which behaved this way, and these samples (custom_robot_window, custom_robot_window_simple) are obviously designed to be shown in a narrow panel.  

However lately in R2022b robot windows always open in a separate browser window, which may be very inconvenient sometimes.  I thought I was doing something wrong in WeBots config or code,  but apparently  WbRobotWindow::openOnWebBrowser()  just always starts an external browser.   No sign of using Qt WebEngine or similar embedded component, as one might expect to see there?   

I'm a fairly new WeBots user and don't know its history well, so could somebody please explain why a documented feature has disappeared from WeBots?  Is the change documented somewhere?

**Comment by stefaniapedrazzi** [comment]
Hi,

since R2022b robot windows are running on standard web browsers and we removed the dependency on the Qt WebEngine module.
This change is an enhancement introducing the new functionalities allowing to run robot windows on a different machine than the one where Webots is executed.
The change is documented in the ChangeLog: https://github.com/cyberbotics/webots/blob/R2022b/docs/reference/changelog-r2022.md?plain=1#L16

Could you link us to the documentation part that is still referring to the robot windows would show up in a special panel in Webots UI?
We may have missed it when updating the documentation.
Thank you!

**Reply by amrytov** [reply]
Hi, 

It's probably too late to discuss, and perhaps there have been some good technical reasons for getting rid of Qt WebEngine, but to me replacing more convenient with much less convenient doesn't look like an enhancement at all.  

I'm developing education scenes in WeBots, where children have to program a robotic system to perform certain simple tasks. The children are young and they don't know WeBots very well, so any extra menu command they have to use is a pain to explain. Often the scenes must include some simple GUI for controlling the robot.   One example is a scene where they must enter text commands to control a robot in a maze.  The **perfect** solution would be to have a way to add GUI controls right into the 3D scene window,  e.g. inside a Billboard node.  The "next best" solution would be to have a GUI panel right next to the 3D scene, within WeBots window.  That was how it worked before the "enhancement".  Even so, one significant disadvantage was that the user has to know how to call up the GUI, it would not appear automatically at a specific position.     

Now with this browser "enhancement",  the GUI page pops up as a separate browser window, overlapping Webots window, they fight for screen space and it is impossible to control the exact size and layout.  I do understand that using a separate browser window does make sense when you are dong complicated things on a dev dual-monitor system in a research lab.   But when all you need is a text box or a couple of buttons, as close to your robot as possible, automatically visible in a fixed position for users who might have trouble finding the right menu to show them,  popping up the GUI in a separate browser is a pure nuisance.

Please-please, consider returning the embedded HTML, or any other sort of basic embedded UI as an option, maybe even make them "more embeddable" by placing at controllable position on the scene, as is done with camera overlays or the Billboard.

**Reply by omichel** [reply]
We do understand your concern on this question. We actually got rid of the dependency on Qt WebEngine because it is not available on the Windows/MSYS2 platform we are using to build Webots. Instead we had to use Qt WbView which not fully compatible with WebEngine and was causing many problems and discrepancies across platforms. Moreover Qt WebEgine is a huge package which was making the Webots installer way bigger than it is today.

Nevertheless we are taking your concern and use-case very seriously and are thinking about various possible solutions:
1. Embed a browser window inside Webots: it might be possible to launch a Firefox or Chrome window inside a Qt sub-window.
2. Improve the Display device to allow for simple user interactions, e.g., click on a button, press keys, etc.
3. Develop a simpler HTML view allowing users to create a mini HTML page. This would however require that we find or develop a lightweight HTML rendering library.

**Reply by omichel** [reply]
We just had a team discussion about this issue and we came out with some fairly elegant solution: we could allow a Robot to be a child of a Billboard node. Obviously, this robot should not have any physics and should typically be a supervisor. It could contain in particular a Display device allowing to display any information to the user using the Display API. Additionally, since the Supervisor API has a mouse monitoring capability, it should be possible to detect mouse clicks and even mouse motion on the Display which will allow user to develop simple widgets, like buttons, check boxes, etc.

Moreover allowing Robots inside Billboards should be pretty easy to implement.

@amrytov, would that resolve the problem for your use-case?

**Reply by amrytov** [reply]
Thank you very much, that would be an extremely useful new feature.  

Although still doesn't quite solve the problem of providing simple input UI.  Perhaps something like https://doc.qt.io/qt-6/qtuitools-module.html  could be used.

**Reply by omichel** [reply]
I believe that it should be possible to build a simple input UI with a Display device in the Billboard by using the supervisor mouse and keyboard APIs. I don't like QtUITool as it introduces a new format that fulfills the same requirements as the HTML robot windows.

**Reply by omichel** [reply]
I discovered today that there might be a fairly simple possibility to re-introduce the HTML robot windows inside Webots by using Qt6 for MSVC and MSYS2 UCRT64, see #5683 for details.
